kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/punctuators/AbstractThrottledPunctuator.java

```diff
@@ -0,0 +1,97 @@
+package org.hypertrace.core.kafkastreams.framework.punctuators;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.Punctuator;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.hypertrace.core.kafkastreams.framework.punctuators.action.TaskResult;
+
+@Slf4j
+public abstract class AbstractThrottledPunctuator<T> implements Punctuator {
+  private final Clock clock;
+  private final KeyValueStore<Long, ArrayList<T>> objectStore;
```

Comment:
Why are we enforcing the need for a KeyValueStore(state store) here?
All we need is a data structure that has data keyed by some time and a way to iterate over them in some order.
If we just expose our custom interface for it, it would be more flexible and we are not forcing the usage of a state store that way.

Reply:
We don't want to lose data on restarts as it will make tasks scheduled inconsistent, hence we enforced this. We can create interface with range, get, put, delete, but not sure if that change will be useful given we are bound to use this punctuator only with Kafka Streams and more often than not we'd be using state stores rather than other mechanisms

Reply:
> We don't want to lose data on restarts as it will make tasks scheduled inconsistent, hence we enforced this

But, thats the case even now. Using a punctuator doesn't mandate using a state store. There may be other stores(depending on the app logic) that may help in rescheduling the punctuators.

Reply:
We have only one punctuator. And this punctuator is stateless. Currently based on app logic of different state stores we have multiple stateful punctuators. 

If we want invocation tasks which are spread on random points in future time, and should be resilient to failures. I think we need a state store based on timestamp as key if we want to make use of range query on time

Reply:
Update, resolved offline: For few use cases we don't need resilience across restarts and tasks can be lost over restarts.. But even that could be supported via using KeyValue store as interface. We just need to provide a in memory store with logging and caching disabled. The control is still in application, so we should be good with KeyValue store as interface